### ISSUE TITLE
opencv3 compilation error

### ISSUE THREAD

**Issue by remia** [Homebrew/homebrew-science#4830]
### Indicate troubleshooting steps taken

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [x] Checked that a similar issue hasn't already been filed?
- [ ] Looked up the software bug tracker if reporting a build error?
- [ ] If you're seeing permission errors tried running `sudo chown -R $(whoami) $(brew --prefix)`?

### Bug reports / build errors:

> /tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/python/src2/cv2.cpp:6:10: fatal error: 'Python.h' file not found

#### Brief summary of issue:

The issue occurs only when installing opencv3 with python3 support, the error seems to indicate a missing header and/or header directory compilation setting. 

The complete command invoking clang++ is : 

`cd /tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/macbuild/modules/python3 && /usr/local/Homebrew/Library/Homebrew/shims/super/clang++   -D__OPENCV_BUILD=1 -Dopencv_python3_EXPORTS -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/Accelerate -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/macbuild/3rdparty/ippicv/ippicv_osx/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/macbuild -isystem /usr/local/include/eigen3 -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/python/python3/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/python/python3/src -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/macbuild/modules/python3 -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/core/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/flann/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/imgproc/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/ml/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/photo/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/video/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/imgcodecs/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/shape/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/videoio/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/highgui/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/objdetect/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/superres/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/features2d/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/calib3d/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/stitching/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/videostab/include -I/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/include/python3.5m -I/usr/local/lib/python3.5/site-packages/numpy/core/include -I/tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/python/src2  -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wno-narrowing -Wno-delete-non-virtual-dtor -Wno-unnamed-type-template-args -Wno-comment -fdiagnostics-show-option -Wno-long-long -Qunused-arguments -Wno-semicolon-before-method-body -fno-omit-frame-pointer -msse -msse2 -mavx -fvisibility=hidden -fvisibility-inlines-hidden -mf16c -Wno-unused-function -Wno-overloaded-virtual -Wno-unused-private-field -DNDEBUG  -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk -fPIC   -o CMakeFiles/opencv_python3.dir/__/src2/cv2.cpp.o -c /tmp/opencv3-20170108-86584-18679kx/opencv-3.2.0/modules/python/src2/cv2.cpp`

We can see the following header directory settings in the command above : 
`-I/usr/local/Cellar/python3/3.5.2_3/Frameworks/Python.framework/Versions/3.5/include/python3.5m`

And indeed Python.h is present in this directory with correct permissions, using brew -vd option to keep temporary file I tried the exact same compilation command after the brew install failure and it works fine.

#### The complete command that you issued:

`brew reinstall opencv3 -vd --with-python3 --with-qt5 --with-ffmpeg --with-opengl`

#### Gist with output of `brew doctor`:

```
brew doctor
Your system is ready to brew.
```

#### Output of `brew config`:

HOMEBREW_VERSION: 1.1.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 619791e83d2781dca1b675e20249a8aebe085c7c
Last commit: 5 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 870f56aa67fb0f0ad6ac14085844149c1cf81d06
Core tap last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: octa-core 64-bit sandybridge
Homebrew Ruby: 2.0.0-p648
Clang: 8.0 build 800
Git: 2.10.2 => /usr/local/bin/git
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.12_2/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: N/A
macOS: 10.12.2-x86_64
Xcode: 8.2.1
CLT: 8.2.0.0.1.1480973914
X11: N/A

### Formula Requests:

**The quickest way to get a new formula into Homebrew is to file a [Pull Request](https://github.com/Homebrew/homebrew-science/pulls)**.

We will consider your request but it may be closed if it's something we're not actively planning to work on.

**Comment by remia** [comment]
Solved the issue by editing the formula `brew edit opencv3` and adding `env :std`: 

```
class Opencv3 < Formula
  desc "Open source computer vision library, version 3"
  homepage "http://opencv.org/"
  env :std
```

I have no idea why it works as I'm not familiar at all with homebrew formulas

**Comment by dev4dev** [comment]
@remia did it build module for python 3? because I get p2 version only =\

**Comment by remia** [comment]
With `--with-python3` flag yes it builds the python3 lib and I just pasted it in the python3 site-packages.
Again, not sure what the `env: std does` in this case, the problem was perhaps due to incorrect setup / environment / ... on my system (`brew doctor` repoted nothing though).

**Comment by dev4dev** [comment]
meh, it doesn't work for me after they updated it to 3.2 =(

```
/usr/local/Cellar/opencv3/3.2.0
❯❯❯ find . -iname '*python*'
./lib/python2.7
```

**Comment by remia** [comment]
You should maybe check that the commands brew use to discover your python3 environment are working correctly :

```
if build.with? "python3"
      py3_config = `python3-config --configdir`.chomp
      py3_include = `python3 -c "import distutils.sysconfig as s; print(s.get_python_inc())"`.chomp
      py3_version = Language::Python.major_minor_version "python3"
      args << "-DPYTHON3_LIBRARY=#{py3_config}/libpython#{py3_version}.#{dylib}"
      args << "-DPYTHON3_INCLUDE_DIR=#{py3_include}"
end
```

**Comment by dev4dev** [comment]
@remia they return correct paths, but still no luck

**Comment by CarlosRDomin** [comment]
I was having the same issue on my new MacBookPro (OS X Sierra): even with the flag `--with-python3` set, `cmake` would not be able to find my Python3 include&lib folders. Running the installation with `--verbose`, I'd actually get the following warning/message in the `cmake` stage:

> _brew install opencv3 [...] --with-python3 --verbose_
[...]
tar xzf /Users/MYUSERNAME/Library/Caches/Homebrew/opencv3--contrib-3.2.0.tar.gz
Failed to import the site module
Your PYTHONPATH points to a site-packages dir for Python 2.x but you are running Python 3.x!
     **PYTHONPATH is currently: "/usr/local/lib/python2.7/site-packages"**
     You should `unset PYTHONPATH` to fix this.
Failed to import the site module
Your PYTHONPATH points to a site-packages dir for Python 2.x but you are running Python 3.x!
     PYTHONPATH is currently: "/usr/local/lib/python2.7/site-packages"
     You should `unset PYTHONPATH` to fix this.

I fixed it by adding the line `ENV["PYTHONPATH"] = lib/"python3.6/site-packages"` (my Python3 version is 3.6, replace with whichever you have) immediately after the line `if build.with? "python3"`.

Hope it helps!

**Comment by dev4dev** [comment]
@CarlosRDomin you are the HERO!

**Comment by iMichka** [comment]
Hi. Thanks for the investigation.

Would you mind opening a pull request with you changes?
See http://docs.brew.sh/How-To-Open-a-Homebrew-Pull-Request.html for the instructions.

Thanks

**Comment by CarlosRDomin** [comment]
Glad it worked @dev4dev! :)
@iMichka Sure, opening a pull request now. I actually tried to do it yesterday (before reading your comment) but the link I found on google was broken, thanks for the correct link!

**Comment by sononix** [comment]
Can these changes get pushed? - currently on mac sierra homebrew install using python3/opencv is breaking 

I ended up using 
ENV["PYTHONPATH"] = "/usr/local/lib/python3.6/site-packages"

**Comment by CarlosRDomin** [comment]
@sononix I opened a PR a month and half ago (#4885), but it never got merged :( I think they weren't sure why PYTHONPATH needed to be unset (or manually set to your actual Python3 folder)

**Comment by iMichka** [comment]
The thing is that as long as we don't understand exactly what is going on, we are reluctant to merge this. You can modify your formula locally to get this fixed, so there is an unofficial workaround.

We need more time and more help from to community to investigate this in detail though.

**Comment by iMichka** [comment]
Fixed by #5424 . Thanks to @CarlosRDomin for finding this out.
